### PR TITLE
bpo-38249: Expand Py_UNREACHABLE() to __builtin_unreachable() in the release mode.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue38249

### CODE REVIEW

File: Include/pymacro.h [hunk 1]
```diff
@@ -118,6 +118,8 @@
         "We've reached an unreachable state. Anything is possible.\n" \
         "The limits were in our heads all along. Follow your dreams.\n" \
         "https://xkcd.com/2200")
+#elif defined(__GNUC__) || defined(__clang__)
```
Comment (ammaraskar) [inline Include/pymacro.h]: For MSVC `defined(_MSC_VER)` you can use `__assume(0)`

https://docs.microsoft.com/en-us/cpp/intrinsics/assume?view=vs-2019

File: Include/pymacro.h [hunk 2]
```diff
@@ -118,6 +118,10 @@
         "We've reached an unreachable state. Anything is possible.\n" \
         "The limits were in our heads all along. Follow your dreams.\n" \
         "https://xkcd.com/2200")
+#elif defined(__GNUC__) || defined(__clang__) || defined(__INTEL_COMPILER)
+#define Py_UNREACHABLE() __builtin_unreachable()
```
Comment (vstinner) [inline Include/pymacro.h]: nitpick: would you mind to add two spaces indentation to ease readability for all `#define Py_UNREACHABLE()`? Example:

```suggestion
#  define Py_UNREACHABLE() __builtin_unreachable()
```

File: Modules/_tracemalloc.c [hunk 3]
```diff
@@ -712,7 +712,7 @@ tracemalloc_realloc(void *ctx, void *ptr, size_t new_size)
 
                The GIL and the table lock ensures that only one thread is
                allocating memory. */
-            Py_UNREACHABLE();
+            Py_FatalError("Memory allocation failed.");
```
Comment (vstinner) [inline Modules/_tracemalloc.c]:
```suggestion
            Py_FatalError("tracemalloc_realloc() failed to allocate a trace");
```

File: Python/pytime.c [hunk 4]
```diff
@@ -746,7 +746,7 @@ _PyTime_GetSystemClock(void)
     _PyTime_t t;
     if (pygettimeofday(&t, NULL, 0) < 0) {
         /* should not happen, _PyTime_Init() checked the clock at startup */
-        Py_UNREACHABLE();
+        Py_FatalError("pygettimeofday() failed.");
```
Comment (vstinner) [inline Python/pytime.c]:
```suggestion
        Py_FatalError("pygettimeofday() failed");
```

File: Python/pytime.c [hunk 5]
```diff
@@ -775,8 +775,7 @@ pymonotonic(_PyTime_t *tp, _Py_clock_info_t *info, int raise)
             _PyTime_overflow();
             return -1;
         }
-        /* Hello, time traveler! */
-        Py_UNREACHABLE();
+        Py_FatalError("Hello, time traveler!");
```
Comment (vstinner) [inline Python/pytime.c]:
```suggestion
        Py_FatalError("pymonotonic: integer overflow");
```

File: Python/pytime.c [hunk 6]
```diff
@@ -1019,7 +1018,7 @@ _PyTime_GetPerfCounter(void)
 {
     _PyTime_t t;
     if (_PyTime_GetPerfCounterWithInfo(&t, NULL)) {
-        Py_UNREACHABLE();
+        Py_FatalError("_PyTime_GetPerfCounterWithInfo() failed.");
```
Comment (vstinner) [inline Python/pytime.c]:
```suggestion
        Py_FatalError("_PyTime_GetPerfCounterWithInfo() failed");
```

File: Python/pytime.c [hunk 7]
```diff
@@ -918,7 +917,7 @@ _PyTime_GetMonotonicClock(void)
     if (pymonotonic(&t, NULL, 0) < 0) {
         /* should not happen, _PyTime_Init() checked that monotonic clock at
            startup */
-        Py_UNREACHABLE();
+        Py_FatalError("pymonotonic() failed.");
```
Comment (vstinner) [inline Python/pytime.c]:
```suggestion
        Py_FatalError("pymonotonic() failed");
```

**Comment by serhiy-storchaka** [comment]
Thank you for your suggestions Victor and Ammar!

**Review by vstinner — APPROVED** [review]
LGTM.

With the updated doc, it is easier for me to tell that the PR is consistent :-)

**Pull request merged**